### PR TITLE
[Quant] [PT2] Enable Inplace Dropout in _move_exported_model_to_eval

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114725
* #114547

**Summary**
Enable Inplace Dropout replacement in `_move_exported_model_to_eval`

**Test Plan**
```
python -u -m pytest -s -v test_quantize_pt2e.py -k test_move_exported_model_to_eval
```

### CODE REVIEW

File: test/quantization/pt2e/test_quantize_pt2e.py [hunk 1]
```diff
@@ -1607,34 +1607,44 @@ def test_embedding_conv_linear_quantization(self):
...
-        self.assertTrue(dropout_node.args[2])
-
-        # Do the subgraph rewriting
-        torch.ao.quantization.move_exported_model_to_eval(m)
-
-        # Assert that dropout op is now replaced with a clone op
-        targets = [n.target for n in m.graph.nodes]
-        self.assertTrue(torch.ops.aten.clone.default in targets)
-        self.assertTrue(torch.ops.aten.native_dropout.default not in targets)
+        for inplace in [True, False]:
```
Comment (andrewor14) [inline test/quantization/pt2e/test_quantize_pt2e.py:1611]: Do you mind refactoring this into a helper function so we have something like:

```
def test_move_exported_model_to_eval(self):
    self._test_move_exported_model_to_eval_dropout(inplace=True)
    self._test_move_exported_model_to_eval_dropout(inplace=False)
```
This will make it easier for us to tell which case failed if things break in the future

**Review by andrewor14 — APPROVED** [review]
Looks good. Just left a minor comment related to the tests

File: torch/ao/quantization/pt2e/eval_utils.py [hunk 2]
```diff
@@ -21,26 +21,28 @@ def _replace_dropout_for_eval(m: torch.fx.GraphModule):
...
     m.recompile()
 
-    def dropout_train(x):
-        return F.dropout(x, p=0.5, training=True)
+    for inplace in [False, True]:
 
-    def dropout_eval(x):
-        return F.dropout(x, p=0.5, training=False)
+        def dropout_train(x):
+            return F.dropout(x, p=0.5, training=True, inplace=inplace)
```
Comment (jerryzh168) [inline torch/ao/quantization/pt2e/eval_utils.py:27]: can we use ignore_literals just for `inplace`?

**Review by jgong5 — APPROVED** [review]
A separate but related question: will there be a functionalization pass on the pre-autograd graph so that we don't have to handle inplace variants in the future?

File: test/quantization/pt2e/test_quantize_pt2e.py (hunk 1 above)
Comment (leslie-fang-intel) [reply test/quantization/pt2e/test_quantize_pt2e.py:1611]: Thanks for the suggestions and modified.

File: torch/ao/quantization/pt2e/eval_utils.py (hunk 2 above)
Comment (leslie-fang-intel) [reply torch/ao/quantization/pt2e/eval_utils.py:27]: Looks like `ignore_literals` already enabled in `replace_pattern_with_filters`, but it still failed to match inplace dropout per the result of UT in this PR.
https://github.com/pytorch/pytorch/blob/34ea0a2bdc7d17f5ca0e895590159e5fd36fffa0/torch/ao/quantization/pt2e/eval_utils.py#L36-L43

**Comment by leslie-fang-intel** [comment]
> A separate but related question: will there be a functionalization pass on the pre-autograd graph so that we don't have to handle inplace variants in the future?

I think it also needed to the task we discussed here: https://github.com/pytorch/pytorch/pull/114540#discussion_r1406484020 to switch using `SubgraphMatcherWithNameNodeMap` for annotation.

File: torch/ao/quantization/pt2e/eval_utils.py (hunk 2 above)
Comment (andrewor14) [reply torch/ao/quantization/pt2e/eval_utils.py:27]: @jerryzh168 Good catch. I think this is actually wrong right now. Since we do `ignore_literals=True` we actually won't match `inplace` correctly.

@leslie-fang-intel It seems we need a match filter to make sure `inplace` is actually the value we expect. Can you add that in a follow up PR? I also don't understand how the existing test passed.

Comment (leslie-fang-intel) [reply torch/ao/quantization/pt2e/eval_utils.py:27]: > @leslie-fang-intel It seems we need a match filter to make sure inplace is actually the value we expect. Can you add that in a follow up PR? I also don't understand how the existing test passed.

Since there is actually no `inplace` parameter for the dropout aten operator, not sure whether the `match filter` is still needed here. @andrewor14

Comment (andrewor14) [reply torch/ao/quantization/pt2e/eval_utils.py:27]: I see, it's because the decomposed aten ops don't have the `inplace` flag, so matching still works. Thanks for explaining.
